_posts/2016-08-02-should-i-use-a-spy-a-double-or-an-instance-double.markdown

```diff
@@ -0,0 +1,108 @@
+instead of actual objects, which if you're using ActiveRecord,
+may result in hitting the database and slowing the suite down. The
+latest release of the [rspec-mocks](https://github.com/rspec/rspec-mocks)
+library bundled with RSpec 3 includes at least three different ways to implement
+a mock object.
+
+We'll look at some of the differences in each by looking at some examples. First
+off, the `spy`:
+
+```ruby
+[1] pry(main)> require 'rspec/mocks/standalone'
+=> true
+[2] pry(main)> user_spy = spy(User)
+=> #<Double User>
+[3] pry(main)> spy.whatever_method
+=> #<Double (anonymous)>
+```
+
+<!--more-->
+
+The `spy` accepts any method calls and always returns itself unless specified.
+If you need the mock object to raise if it receives an unexpected method call,
+you can use a `double` instead:
+
+```ruby
+[4] pry(main)> user_double = double(User)
+=> #<Double User>
+[5] pry(main)> user_double.whatever_method
+RSpec::Mocks::MockExpectationError: #<Double User> received unexpected message :whatever_method with (no args)
+from /Users/mauro-oto/.rvm/gems/ruby-2.2.1@carbide/gems/rspec-support-3.5.0/lib/rspec/support.rb:87:in block in <module:Support>
+```
+
+If you haven't specified that the double can receive a given method, in the case
+above `whatever_method`, it'll raise an exception. You can explicitly tell the
+double that it can receive such a method and its return value like this:
+
+```ruby
+[6] pry(main)> user_double = double(User, whatever_method: nil)
+=> #<Double User>
+[7] pry(main)> user_double.whatever_method
+=> nil
+[8] pry(main)> user_double.some_method
+RSpec::Mocks::MockExpectationError: #<Double User> received unexpected message :some_method with (no args)
+from /Users/mauro-oto/.rvm/gems/ruby-2.2.1@carbide/gems/rspec-support-3.5.0/lib/rspec/support.rb:87:in block in <module:Support>
+```
+
+This way, `whatever_method` can be called and `nil` will be returned, which is
+the return value we specified, and any other method calls will fail if we
+haven't specified them, like `some_method`.
+
+If we want to have even more control over what happens with our mock object, and
+disallow arbitrary method creation like `whatever_method` or `some_method`, we
+can use a verifying double, which exist since RSpec 3 as `instance_double`:
+
+```ruby
+[9] pry(main)> user_verifiable = instance_double(User, whatever_method: nil)
+RSpec::Mocks::MockExpectationError: the User class does not implement the instance method: whatever_method
+from /Users/mauro-oto/.rvm/gems/ruby-2.2.1@carbide/gems/rspec-support-3.5.0/lib/rspec/support.rb:87:in block in <module:Support>
+```
+
+If we try to declare a method which is not implemented by the class of the
+mocked instance, it will raise an exception. If we decide to use mock objects in
+our tests, instance_doubles provides us with a bit more confidence in our tests
+than if we were using spies or regular doubles.
+
+instance_doubles have a slightly worse performance than doubles or spies, as is
```

Comment:
@mauro-oto I'd say: 

```
The performance of `instance_double` is slightly worse than `double` or `spy` because it is more complex. The difference between `instance_double` and using a real object (using the database) is not that significant: 
```